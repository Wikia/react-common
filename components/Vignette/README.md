Default mode:
```js
<Vignette
	image="https://vignette.wikia.nocookie.net/541b323b-a3c2-4d4f-be1f-500cf13573b8"
	width="100"
	height="200"
/>
```

Other methods:
```js
<div>
	<Vignette
		image="https://vignette.wikia.nocookie.net/541b323b-a3c2-4d4f-be1f-500cf13573b8"
		width="100"
		height="80"
		method="scale"
	/>
	<Vignette
		image="https://vignette.wikia.nocookie.net/541b323b-a3c2-4d4f-be1f-500cf13573b8"
		width="100"
		height="80"
		method="thumbnail"
	/>
	<Vignette
		image="https://vignette.wikia.nocookie.net/541b323b-a3c2-4d4f-be1f-500cf13573b8"
		width="100"
		height="80"
		method="top-crop"
	/>
</div>
```
