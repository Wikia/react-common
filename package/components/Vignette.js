'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var PropTypes = _interopDefault(require('prop-types'));

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};

  var target = _objectWithoutPropertiesLoose(source, excluded);

  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

function getUuid(urlOrUuid) {
  var matches = urlOrUuid.match(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/);

  if (matches) {
    // UUID found in `urlOrUuid`
    return matches[0];
  }

  return false;
}

function getVignetteParamsTopCrop(width, height) {
  return "/top-crop/width/".concat(width, "/height/").concat(height);
}

function getVignetteParamsThumbnail(width, height, allowUpscaling) {
  if (allowUpscaling) {
    return "/thumbnail/width/".concat(width, "/height/").concat(height);
  }

  return "/thumbnail-down/width/".concat(width, "/height/").concat(height);
}

function getVignetteParamsScale(width, height, allowUpscaling) {
  if (width) {
    if (allowUpscaling) {
      return "/scale-to-width/".concat(width);
    }

    return "/scale-to-width-down/".concat(width);
  }

  if (height) {
    return "/scale-to-height-down/".concat(height);
  }

  return '';
}

function getVignetteParams(_ref) {
  var width = _ref.width,
      height = _ref.height,
      method = _ref.method,
      allowUpscaling = _ref.allowUpscaling;

  switch (method) {
    case 'top-crop':
      return getVignetteParamsTopCrop(width, height);

    case 'thumbnail':
      return getVignetteParamsThumbnail(width, height, allowUpscaling);

    case 'scale':
      return getVignetteParamsScale(width, height, allowUpscaling);

    default:
      // auto
      if (width && height) {
        return getVignetteParamsThumbnail(width, height, allowUpscaling);
      }

      return getVignetteParamsScale(width, height, allowUpscaling);
  }
}

/**
 * Vignette helper for getting scaled/resized images from Static Image Assets service.
 *
 * Works for any URL (non-vignette ones won't be resized) and for UUIDs.
 * The `mode` along with `width`, `height` and '`allowUpscaling` will dictate if
 * the final image will be scaled, resized or cropped.
 */

var Vignette = function Vignette(_ref) {
  var allowUpscaling = _ref.allowUpscaling,
      alt = _ref.alt,
      className = _ref.className,
      height = _ref.height,
      image = _ref.image,
      method = _ref.method,
      width = _ref.width,
      rest = _objectWithoutProperties(_ref, ["allowUpscaling", "alt", "className", "height", "image", "method", "width"]);

  var imageUrlOrUuid = image.replace('//static.wikia.nocookie.net/', '//vignette.wikia.nocookie.net/');

  if (imageUrlOrUuid.indexOf('vignette.wikia.nocookie.net') !== -1) {
    var uuid = getUuid(imageUrlOrUuid);

    if (uuid) {
      var params = getVignetteParams({
        width: width,
        height: height,
        method: method,
        allowUpscaling: allowUpscaling
      });
      imageUrlOrUuid = "https://vignette.wikia.nocookie.net/".concat(uuid).concat(params);
    }
  }

  return React.createElement("img", _extends({
    className: className,
    src: imageUrlOrUuid,
    alt: alt
  }, rest));
};

Vignette.propTypes = {
  /** Do we want to upscale image if needed? */
  allowUpscaling: PropTypes.bool,

  /** Alt text */
  alt: PropTypes.string,

  /** Additional class name */
  className: PropTypes.string,

  /** Desired image height */
  height: PropTypes.number,

  /** Either an URL to image or UUID. */
  image: PropTypes.string.isRequired,

  /** Desired image mode */
  method: PropTypes.oneOf(['auto', 'scale', 'thumbnail', 'top-crop']),

  /** Desired image width */
  width: PropTypes.number
};
Vignette.defaultProps = {
  allowUpscaling: false,
  alt: '',
  className: '',
  height: null,
  method: 'auto',
  width: null
};

module.exports = Vignette;
