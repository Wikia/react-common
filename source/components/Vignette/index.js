import React from 'react';
import PropTypes from 'prop-types';

import { getVignetteParams, getUuid } from '../../utils/vignetteHelper';

/**
 * Vignette helper for getting scaled/resized images from Static Image Assets service.
 *
 * Works for any URL (non-vignette ones won't be resized) and for UUIDs.
 * The `mode` along with `width`, `height` and '`allowUpscaling` will dictate if
 * the final image will be scaled, resized or cropped.
 */
const Vignette = ({
    allowUpscaling,
    alt,
    className,
    height,
    image,
    method,
    width,
    ...rest
}) => {
    let imageUrlOrUuid = image.replace('//static.wikia.nocookie.net/', '//vignette.wikia.nocookie.net/');

    if (imageUrlOrUuid.indexOf('vignette.wikia.nocookie.net') !== -1) {
        const uuid = getUuid(imageUrlOrUuid);

        if (uuid) {
            const params = getVignetteParams({
                width, height, method, allowUpscaling,
            });
            imageUrlOrUuid = `https://vignette.wikia.nocookie.net/${uuid}${params}`;
        }
    }

    return <img className={className} src={imageUrlOrUuid} alt={alt} {...rest} />;
};

Vignette.propTypes = {
    /** Do we want to upscale image if needed? */
    allowUpscaling: PropTypes.bool,
    /** Alt text */
    alt: PropTypes.string,
    /** Additional class name */
    className: PropTypes.string,
    /** Desired image height */
    height: PropTypes.number,
    /** Either an URL to image or UUID. */
    image: PropTypes.string.isRequired,
    /** Desired image mode */
    method: PropTypes.oneOf([
        'auto',
        'scale',
        'thumbnail',
        'top-crop',
    ]),
    /** Desired image width */
    width: PropTypes.number,
};

Vignette.defaultProps = {
    allowUpscaling: false,
    alt: '',
    className: '',
    height: null,
    method: 'auto',
    width: null,
};

export default Vignette;
