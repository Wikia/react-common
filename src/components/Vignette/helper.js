export function getUuid(urlOrUuid) {
  const matches = urlOrUuid.match(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/);

  if (matches) {
    // UUID found in `urlOrUuid`
    return matches[0];
  }

  return false;
}

function getVignetteParamsTopCrop(width, height) {
  return `/top-crop/width/${width}/height/${height}`;
}

function getVignetteParamsThumbnail(width, height, allowUpscaling) {
  if (allowUpscaling) {
    return `/thumbnail/width/${width}/height/${height}`;
  }
  return `/thumbnail-down/width/${width}/height/${height}`;
}

function getVignetteParamsScale(width, height, allowUpscaling) {
  if (width) {
    if (allowUpscaling) {
      return `/scale-to-width/${width}`;
    }
    return `/scale-to-width-down/${width}`;
  }

  if (height) {
    return `/scale-to-height-down/${height}`;
  }

  return '';
}

export function getVignetteParams({width, height, method, allowUpscaling}) {
  switch (method) {
    case 'top-crop':
      return getVignetteParamsTopCrop(width, height);
    case 'thumbnail':
      return getVignetteParamsThumbnail(width, height, allowUpscaling);
    case 'scale':
      return getVignetteParamsScale(width, height, allowUpscaling);
    default:
      // auto
      if (width && height) {
        return getVignetteParamsThumbnail(width, height, allowUpscaling);
      }
      return getVignetteParamsScale(width, height, allowUpscaling);
  }
}
