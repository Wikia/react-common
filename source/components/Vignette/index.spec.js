import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Vignette from './index';

test('Vignette renders correctly with default values', () => {
    const component = renderer.create(
        <Vignette
            image="https://vignette.wikia.nocookie.net/541b323b-a3c2-4d4f-be1f-500cf13573b8"
        />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Vignette renders correctly with non-vignette image', () => {
    const component = renderer.create(
        <Vignette
            image="https://example.com/vimage.jpeg"
        />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Vignette renders correctly with broken vignette image', () => {
    const component = renderer.create(
        <Vignette
            image="https://vignette.wikia.nocookie.net/path/to/image.jpeg"
        />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Vignette renders plain image tag when no width and height present', () => {
    const image = 'https://vignette.wikia.nocookie.net/541b323b-a3c2-4d4f-be1f-500cf13573b8';
    const component = shallow(
        <Vignette image={image} />
    );
    expect(component.find('img').prop('src')).toEqual(image);
});
