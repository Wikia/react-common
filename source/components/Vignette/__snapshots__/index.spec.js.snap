// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vignette renders correctly with broken vignette image 1`] = `
<img
  alt=""
  className=""
  src="https://vignette.wikia.nocookie.net/path/to/image.jpeg"
/>
`;

exports[`Vignette renders correctly with default values 1`] = `
<img
  alt=""
  className=""
  src="https://vignette.wikia.nocookie.net/541b323b-a3c2-4d4f-be1f-500cf13573b8"
/>
`;

exports[`Vignette renders correctly with non-vignette image 1`] = `
<img
  alt=""
  className=""
  src="https://example.com/vimage.jpeg"
/>
`;
